src/libraries/System.Private.CoreLib/src/System/Text/StringBuilder.cs

```diff
@@ -659,36 +659,58 @@ public StringBuilder Append(char value, int repeatCount)
                 return this;
             }
 
-            // this is where we can check if the repeatCount will put us over m_MaxCapacity
-            // We are doing the check here to prevent the corruption of the StringBuilder.
-            int newLength = Length + repeatCount;
-            if (newLength > m_MaxCapacity || newLength < repeatCount)
+            char[] chunkChars = m_ChunkChars;
+            int chunkLength = m_ChunkLength;
+
+            // Try to fit the whole repeatCount in the current chunk
+            // Use the same check as Span<T>.Slice for 64-bit so it can be folded
+            // Since repeatCount can't be negative, there's no risk for it to overflow on 32 bit
+            if (((nuint)(uint)chunkLength + (nuint)(uint)repeatCount) <= (nuint)(uint)chunkChars.Length)
             {
-                throw new ArgumentOutOfRangeException(nameof(repeatCount), SR.ArgumentOutOfRange_LengthGreaterThanCapacity);
+                chunkChars.AsSpan(chunkLength, repeatCount).Fill(value);
+                m_ChunkLength += repeatCount;
             }
-
-            int index = m_ChunkLength;
-            while (repeatCount > 0)
+            else
             {
-                if (index < m_ChunkChars.Length)
-                {
-                    m_ChunkChars[index++] = value;
-                    --repeatCount;
-                }
-                else
-                {
-                    m_ChunkLength = index;
-                    ExpandByABlock(repeatCount);
-                    Debug.Assert(m_ChunkLength == 0);
-                    index = 0;
-                }
+                AppendWithExpansion(value, repeatCount);
             }
 
-            m_ChunkLength = index;
             AssertInvariants();
             return this;
         }
 
+        [MethodImpl(MethodImplOptions.NoInlining)]
```

Comment:
JIT does not inline methods that throw, so this attribute is not needed here

@EgorBo please correct me if I am wrong (JIT is changing very rapidly)

```suggestion
```

Reply:
> JIT does not inline methods that throw

Only in case if it's a throw helper (when jit scans a method and see that it never returns). if method has conditions like `if (...) throw` - it's fine. JIT currently never inlines methods with exception handling inside. As for this method - it looks pretty big so I'd be surprised it could be inlined without the attribute